crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -727,117 +728,152 @@ impl<P: RoomDataProvider> TimelineInner<P> {
-                    return None;
+            // Loop through all the indices, in order so we don't decrypt edits
+            // before the event being edited, if both were UTD. Keep track of
+            // index change as UTDs are removed instead of updated.
+            let mut offset = 0;
+            for idx in retry_indices {
+                let idx = idx - offset;
+                let Some(mut event) = retry_one(state.items[idx].clone()).await else {
+                    continue;
                 };
 
-                tracing::Span::current().record("event_id", debug(&remote_event.event_id));
-
-                match decryptor.decrypt_event_impl(&remote_event.original_json).await {
-                    Ok(event) => {
-                        trace!("Successfully decrypted event that previously failed to decrypt");
-                        Some(event)
-                    }
-                    Err(e) => {
-                        info!("Failed to decrypt event after receiving room key: {e}");
-                        None
-                    }
+                event.push_actions = push_rules_context
+                    .as_ref()
+                    .map(|(push_rules, push_context)| {
+                        push_rules.get_actions(&event.event, push_context).to_owned()
+                    })
+                    .unwrap_or_default();
+
+                let result = state
+                    .handle_remote_event(
+                        event.into(),
+                        TimelineItemPosition::Update(idx),
+                        &room_data_provider,
+                        track_read_receipts,
+                    )
+                    .await;
+
+                // If the UTD was removed rather than updated, offset all
+                // subsequent loop iterations.
+                if result.item_removed {
+                    offset += 1;
                 }
             }
-            .instrument(info_span!(
-                "retry_one",
-                session_id = field::Empty,
-                event_id = field::Empty
-            ))
         };
 
-        let mut state = self.state.lock().await;
-
-        // We loop through all the items in the timeline, if we successfully
-        // decrypt a UTD item we either replace it or remove it and update
-        // another one.
-        let mut idx = 0;
-        while let Some(item) = state.items.get(idx) {
-            let Some(mut event) = retry_one(item.clone()).await else {
-                idx += 1;
-                continue;
-            };
-
-            event.push_actions = push_rules_context
-                .as_ref()
-                .map(|(push_rules, push_context)| {
-                    push_rules.get_actions(&event.event, push_context).to_owned()
-                })
-                .unwrap_or_default();
+        // For testing it's good to wait for this to finish
+        #[cfg(test)]
+        do_retry.await;
```

Comment:
`yield_now()` will let the task make *some* progress but I think there's a high chance it yields itself before finishing, so I don't think that's a great solution. We could try though if you feel like this is important.

Also `cfg(test)` only applies to unit tests, not integration tests. So we can still cover the other code path with tests.

Reply:
Please try with `yield_now` or (worst) a timer (`sleep`). If you believe it's too ugly, then let's keep as is.

Reply:
Apparently it wasn't even needed! I'm a little worried about how long it will stay that way, but clearly now we just don't need the different behavior for `cfg(test)´.